src/FreeAge/server/game.cpp

```diff
@@ -335,8 +335,15 @@ void Game::HandlePlaceBuildingFoundationMessage(const QByteArray& msg, PlayerInG
   const char* data = msg.data();
   
   BuildingType type = static_cast<BuildingType>(mango::uload16(data + 3));
-  // TODO: Check whether the player is allowed to build this type of building
-  
+
+  // check whether the player is allowed to build this type of building
+  int max = GetBuildingMaxInstances(type);
+  bool maxReached = max != -1 && player->stats.GetBuildingTypeCount(type) >= max;
```

Comment:
The whole -1 thing is ugly but I prefer, any opinions?

Reply:
What would speak against using `std::numeric_limits<int>::max()`? This would remove the need to do a special case check.